function-controller/pkg/controller/autoscaler/autoscaler.go

```
@@ -0,0 +1,325 @@
+	return func(mt *metricsTotals) int {
+		return a(s(mt))
+	}
+}
+
+func (a *autoScaler) metricsScaler(fn FunctionId) func(mt *metricsTotals) int {
+	return func(mt *metricsTotals) int {
+		var proposedReplicas int
+		if mt.receiveCount == 0 {
+			if mt.transmitCount == 0 {
+				proposedReplicas = 0
+			} else {
+				proposedReplicas = 1 // arbitrary value
+			}
+		} else {
+			proposedReplicas = int(math.Ceil(float64(a.replicas[fn]) * float64(mt.transmitCount) / float64(mt.receiveCount)))
+		}
+		return proposedReplicas
+	}
+}
+
+func (a *autoScaler) limitScalingUp(fn FunctionId) func(proposedReplicas int) int {
+	return func(proposedReplicas int) int {
+		maxReplicas := a.maxReplicas(fn)
+		possibleChange := proposedReplicas != a.replicas[fn]
+		if proposedReplicas > maxReplicas {
+			if possibleChange {
+				log.Printf("Proposing %v should have maxReplicas (%d) instead of %d replicas", fn, maxReplicas, proposedReplicas)
+			}
+			proposedReplicas = maxReplicas
+		}
+		return proposedReplicas
+	}
+}
+
+func (a *autoScaler) limitScalingDown(fn FunctionId) func(proposedReplicas int) int {
+	return func(proposedReplicas int) int {
+		// If zero replicas are proposed *and* there is already at least one replica, check the queue of work to the function.
+		// The queue length is not allowed to initiate scaling up from 0 to 1 as that would confuse rate-based autoscaling.
+		if proposedReplicas == 0 && a.replicas[fn] != 0 {
+			empty, length := a.emptyQueue(fn)
+			if !empty {
+				// There may be work to do, so propose 1 replica instead.
+				log.Printf("Ignoring proposal to scale %v to 0 replicas since queue length is %d", fn.Function, length)
+				proposedReplicas = 1
+			}
+		}
+		return proposedReplicas
+	}
+}
+
+func (a *autoScaler) InformFunctionReplicas(function FunctionId, replicas int) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	a.replicas[function] = replicas
+}
+
+func (a *autoScaler) Close() error {
+	close(a.stop)
```

Is a closed autoscaler supposed to be re-usable? I assume not, but Run() should fail fast (or reset all state)

---

No, it isn't. I've fixed this by making `Close()` set the mutex to `nil` rather than carrying around extra state to track whether the autoscaler has been closed. I don't think we need to provide a graceful failure if a closed autoscaler is reused as that would be a basic bug in the function-controller.